src/aleph/services/p2p/protocol.py

```diff
@@ -29,100 +28,84 @@
-                            try:
-                                self.peers[peer].remove((stream, semaphore))
-                            except ValueError:
-                                pass
-                            LOGGER.debug("Can't request hash...")
-                await asyncio.sleep(0)
-
-            if not len(streams):
-                return
+        read_bytes = await stream.receive_some(MAX_READ_LEN)
+        if read_bytes is None:
+            return
+
+        try:
+            read_string = read_bytes.decode("utf-8")
+            message_json = json.loads(read_string)
+            if message_json["command"] == "hash_content":
+                value = await get_hash_content(
+                    message_json["hash"], use_network=False, timeout=1
+                )
+                if value is not None and value != -1:
+                    result = {
+                        "status": "success",
+                        "hash": message_json["hash"],
+                        "content": base64.encodebytes(value).decode("utf-8"),
+                    }
+                else:
+                    result = {"status": "success", "content": None}
+            elif message_json["command"] == "get_message":
+                result = {"status": "error", "reason": "not implemented"}
+            elif message_json["command"] == "publish_message":
+                result = {"status": "error", "reason": "not implemented"}
+            elif message_json["command"] == "hello":
+                result = {
+                    "status": "success",
+                    "content": {"version": __version__},
+                }
+            else:
+                result = {"status": "error", "reason": "unknown command"}
+            LOGGER.debug(f"received {read_string}")
+        except Exception as e:
+            result = {"status": "error", "reason": repr(e)}
+            LOGGER.exception("Error while reading data")
+
+        await stream.send_all(json.dumps(result).encode("utf-8"))
+
+    async def make_request(self, request_structure: Dict[str, Any]) -> Optional[Dict[str, Any]]:
+        peers = [peer for peer in self.peers]
+        # Randomize the list of peers to contact to distribute the load evenly
+        random.shuffle(peers)
+
+        for peer in peers:
+            stream_info, stream = self.p2p_client.stream_open(peer, (self.PROTOCOL_ID,))
+            msg = json.dumps(request_structure).encode("UTF-8")
+            await stream.send_all(msg)
+
+            response = await stream.receive_some(MAX_READ_LEN)
+            try:
+                value = json.loads(response.decode("UTF-8"))
+            except json.JSONDecodeError:
+                logging.warning("Could not decode response from %s", peer)
+                continue
+
+            if value.get("content") is not None:
+                return value
+
+        logging.info("Could not retrieve content from any peer")
```

Comment:
Should this be a warning ?

Reply:
I don't think so, I don't think we check that the content must exist beforehand?